### DISCUSSION TITLE
error: redefinition of module 'SwiftBridging'

### DISCUSSION THREAD

**Discussion by shikiroot** [Homebrew/discussions#5723] [Getting started] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.4.4
ORIGIN: https://github.com/Homebrew/brew
HEAD: 824efa8836dc226aa92dbbf7404c1b4a66707cca
Last commit: 4 days ago
Core tap JSON: 08 Nov 05:49 UTC
Core cask tap JSON: 08 Nov 05:49 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 3.3.5 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.5/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 16.0.0 build 1600
Git: 2.47.0 => /opt/homebrew/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 15.1-arm64
CLT: 16.1.0.0.1.1729049160
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: No Cask quarantine support available: unknown reason.
```

### Description of issue

Device: M1 Pro MBP
OS: 15.1
While using `brew upgrade <cask>` or `brew upgrade --greedy`, return:
```
/Library/Developer/CommandLineTools/usr/include/swift/module.modulemap:13:8: error: redefinition of module 'SwiftBridging'
11 | //===----------------------------------------------------------------------===//
12 |
13 | module SwiftBridging {
   |        `- error: redefinition of module 'SwiftBridging'
14 |     header "bridging"
15 |

/Library/Developer/CommandLineTools/usr/include/swift/bridging.modulemap:13:8: note: previously defined here
11 | //===----------------------------------------------------------------------===//
12 |
13 | module SwiftBridging {
   |        `- note: previously defined here
14 |     header "bridging"
15 |

/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Foundation.framework/Modules/Foundation.swiftmodule/arm64e-apple-macos.swiftinterface:5:8: error: failed to build module 'CoreFoundation'; this SDK is not supported by the compiler (the SDK is built with 'Apple Swift version 6.0 effective-5.10 (swiftlang-6.0.0.7.43 clang-1600.0.25.3)', while this compiler is 'Apple Swift version 6.0.2 effective-5.10 (swiftlang-6.0.2.1.2 clang-1600.0.26.4)'). Please select a toolchain which matches the SDK.
    3 | // swift-module-flags: -target arm64e-apple-macos15.1 -target-variant arm64e-apple-ios18.1-macabi -enable-objc-interop -autolink-force-load -enable-library-evolution -module-link-name swiftFoundation -swift-version 5 -enforce-exclusivity=unchecked -O -library-level api -enable-upcoming-feature InferSendableFromCaptures -enable-bare-slash-regex -user-module-version 3107 -module-name Foundation -package-name "\"FoundationPreview\""
    4 | import Combine
    5 | import CoreFoundation
      |        `- error: failed to build module 'CoreFoundation'; this SDK is not supported by the compiler (the SDK is built with 'Apple Swift version 6.0 effective-5.10 (swiftlang-6.0.0.7.43 clang-1600.0.25.3)', while this compiler is 'Apple Swift version 6.0.2 effective-5.10 (swiftlang-6.0.2.1.2 clang-1600.0.26.4)'). Please select a toolchain which matches the SDK.
    6 | import Darwin
    7 | import Dispatch

/Library/Developer/CommandLineTools/usr/include/swift/module.modulemap: 0:
/Library/Developer/CommandLineTools/usr/include/swift/bridging.modulemap: 0:
/opt/homebrew/Library/Homebrew/cask/utils/quarantine.swift:3:8: error: failed to build module 'Foundation'; this SDK is not supported by the compiler (the SDK is built with 'Apple Swift version 6.0 effective-5.10 (swiftlang-6.0.0.7.43 clang-1600.0.25.3)', while this compiler is 'Apple Swift version 6.0.2 effective-5.10 (swiftlang-6.0.2.1.2 clang-1600.0.26.4)'). Please select a toolchain which matches the SDK.
 1 | #!/usr/bin/swift
 2 |
 3 | import Foundation
   |        `- error: failed to build module 'Foundation'; this SDK is not supported by the compiler (the SDK is built with 'Apple Swift version 6.0 effective-5.10 (swiftlang-6.0.0.7.43 clang-1600.0.25.3)', while this compiler is 'Apple Swift version 6.0.2 effective-5.10 (swiftlang-6.0.2.1.2 clang-1600.0.26.4)'). Please select a toolchain which matches the SDK.
 4 |
 5 | struct SwiftErr: TextOutputStream {
```

Tried `sudo rm -rf /Library/Developer/CommandLineTools/` and `xcode-select --install`, issue still exist.
Use `swift $(brew --repo)/Library/Homebrew/cask/utils/quarantine.swift` return same error.

**Comment by carlocab** [comment]
Seems related to https://github.com/Homebrew/brew/issues/18656, and will likely be fixed by https://github.com/Homebrew/brew/issues/18656#issuecomment-2463847356

**Comment by Bo98** [comment, marked as answer]
We've seen some reports of bugs with `xcode-select --install` where multiple CLT versions get installed together, which breaks things entirely.

If you can, it's worth trying `sudo rm -rf /Library/Developer/CommandLineTools/` and then instead of `xcode-select --install` download "Command Line Tools for Xcode 16.1" from here: https://download.developer.apple.com/Developer_Tools/Command_Line_Tools_for_Xcode_16.1/Command_Line_Tools_for_Xcode_16.1.dmg (needs a developer account but that's free).

(note for future readers: check https://developer.apple.com/download/all/ if you're reading this a few weeks from now as there may be a newer version - 16.2 will probably be out in December)

**Reply by aaronblair** [reply]
This worked for me. I hope that this swift dependency issue gets fixed long term. Cheers!

**Reply by shikiroot** [reply]
Does not work for me. Tried on a new MBP 15.1, install CLT, then install homebrew, still get `Warning: No Cask quarantine support available: unknown reason.` while running `brew doctor` and same error with `swift $(brew --repo)/Library/Homebrew/cask/utils/quarantine.swift`.

**Reply by dlhanson46** [reply]
Worked for me -- my system is ready to brew again. Thank you!

**Reply by frankievalentine** [reply]
This works for me thanks!

**Reply by hankipanky** [reply]
Worked for me, too, using "Command Line Tools for Xcode 16.2" from here: https://download.developer.apple.com/Developer_Tools/Command_Line_Tools_for_Xcode_16.2/Command_Line_Tools_for_Xcode_16.2.dmg

**Reply by tony-sol** [reply]
not worked for me

UPD: installing Xcode was a fix in my case

**Reply by lzandman** [reply]
This worked for me too!

**Reply by g1dd5** [reply]
What can you do if you don't have an Apple Developer account, and so none of those links are available?  (I don't have XCode installed, just the Command Line Tools.)

**Reply by lzandman** [reply]
> What can you do if you don't have an Apple Developer account

Well, just create an Apple Developer Account. It should be free.

**Reply by tony-sol** [reply]
> What can you do if you don't have an Apple Developer account, and so none of those links are available?  (I don't have XCode installed, just the Command Line Tools.)

try to just download XCode from AppStore - it doesn't require a developer account

**Reply by g1dd5** [reply]
Actually, it turns out that you don't need an Apple Developer account to download the Command Line Tools — just an Apple ID. So now HomeBrew is working fine!  \o/  Thanks all!

One minor caveat: it had forgotten the casks that gave errors before. I had to install them again, but that went fine.

(And that raised an unrelated issue: the Puzzles cask is now deprecated in favour of a formula, which can't be put in the Dock (unless you want to remove, find, and re-add it after every upgrade). So I cobbled together a trivial runner app — details on request.)

**Reply by prantlf** [reply]
This still helps! I wonder what's wrong with `xcode-select --install`. [The latest version of the dev tools is 16.2](https://download.developer.apple.com/Developer_Tools/Command_Line_Tools_for_Xcode_16.2/Command_Line_Tools_for_Xcode_16.2.dmg) now.

**Reply by rosingrind** [reply]
`swift $(brew --repo)/Library/Homebrew/cask/utils/quarantine.swift` and `brew doctor` works without warnings on 16.1, but it started to happen again on 16.2; rolling back to 16.1 again helped. Is there any way to get it working on latest 16.2?

**Reply by vikkastaneja** [reply]
I am also seeing the same behavior for 16.2. I don't know how to revert back to 16.1 without developer account, which apple is charging me 99 bucks. Any pointers?

**Reply by gromgit** [reply]
Go to the [main Apple Developer download page](https://developer.apple.com/download/all/). You'll have to login with your Apple ID, but you definitely don't have to fork over USD99.

**Reply by prantlf** [reply]
> Go to the [main Apple Developer download page](https://developer.apple.com/download/all/). You'll have to login with your Apple ID, but you definitely don't have to fork over USD99.

I downloaded the [16.2 command-line dev tools](https://download.developer.apple.com/Developer_Tools/Command_Line_Tools_for_Xcode_16.2/Command_Line_Tools_for_Xcode_16.2.dmg), not the full Xcode, without spending a cent. I had to log in with my Apple account, of course, but that's free.

**Reply by prantlf** [reply]
> `swift $(brew --repo)/Library/Homebrew/cask/utils/quarantine.swift` and `brew doctor` works without warnings on 16.1, but it started to happen again on 16.2; rolling back to 16.1 again helped. Is there any way to get it working on latest 16.2?

Both commands succeed on my machine with 16.2. `swift` is silent and `doctor` complains only about some deprecated packages.

**Reply by dcchambers** [reply]
I was also getting Swift errors when trying to upgrade homebrew casks after upgrading from MacOS 14 Sonoma to 15 Sequoia.

(Re)installing the command line tools via `xcode-select --install` did not work for me, but manually grabbing the 16.2 CLT .dmg from the apple developer website and then installing did fix the issue.

Not sure what's up with the `xcode-select` installer 🤷

**Comment by shikiroot** [comment]
Tried install Xcode from APP Store and remove CLT directory, problem solved.

**Comment by vikkastaneja** [comment]
I think reinstalling with xcode commandline install 16.3, which is causing
the issue.

**Comment by Aternus** [comment]
# Diagnosis
The problem seems to lie with the `xcode-select` installer triggering 15 installation (probably someone forgot to change the build string for 16, hehe).

# Solution
Run `softwareupdate --install "Command Line Tools for Xcode-16.2"` after the first `xcode-select --install` to force the latest version.

# Aside
If you `ls -al /Library/Developer/CommandLineTools/SDKs` you'll see that the default SDK points to 15.2, which is normal, and part of the issue that the CLT installer has which triggers the 15 mixup, hopefully Apple devs will fix the build in the next version.

❗ If you've did the steps correctly you should see only two SDK versions in this folder, 14.5 and 15.2 (as of 2025.02.12).

![Screen Shot 2025-02-12 at 12 07 39](https://github.com/user-attachments/assets/673d2723-3b63-4241-bcf9-c6a1a3d4cc8f)

P.S. - This is the same as manually installing the 16.2 SDK from the website. If you need to have the actual 15 SDK then you need to install a previous version of CLT without the build bug.

**Reply by oytech** [reply]
### Problem
I confirm your findings - `xcode-select --install` installed 2 different versions for me at the same time. It could be seen by running `softwareupdate --history`:
```
macOS Sonoma 14.7.5                                14.7.5     03.05.2025, 20:59:28
Command Line Tools for Xcode                       15.3       03.05.2025, 21:31:34
Command Line Tools for Xcode                       16.2       03.05.2025, 21:31:34
```
Also I had more than 2 SDKs in `/Library/Developer/CommandLineTools/SDKs`.

### Solution
First I removed it all:
```bash
sudo rm -rf /Library/Developer/CommandLineTools/
```

Then I fixed it using trick I found in https://apple.stackexchange.com/questions/107307/how-can-i-install-the-command-line-tools-completely-from-the-command-line:
```bash
touch /tmp/.com.apple.dt.CommandLineTools.installondemand.in-progress
# copy label from softwareupdate --list
softwareupdate --install "Command Line Tools for Xcode-16.2"
rm /tmp/.com.apple.dt.CommandLineTools.installondemand.in-progress
```

Now I have 2 SDKs `MacOSX14.5.sdk`, `MacOSX15.2.sdk`  in `/Library/Developer/CommandLineTools/SDKs` like on your screenshot.

**Reply by FalcoG** [reply]
The solution to this problems seems to be much simpler now with macOS 15.4.1.

The problem was resolved by cleaning the directory with ``sudo rm -rf /Library/Developer/CommandLineTools/`` and then installing xcode with ``xcode-select --install``.

Before:

```
ls -al /Library/Developer/CommandLineTools/SDKs
total 0
drwxr-xr-x  10 root  wheel  320 Apr 10 23:32 .
drwxr-xr-x   5 root  wheel  160 Feb 12 18:44 ..
lrwxr-xr-x   1 root  wheel   14 Apr 10 23:32 MacOSX.sdk -> MacOSX15.4.sdk
drwxr-xr-x   7 root  wheel  224 May 12  2023 MacOSX13.3.sdk
lrwxr-xr-x   1 root  wheel   14 Feb 12 18:45 MacOSX13.sdk -> MacOSX13.3.sdk
drwxr-xr-x   7 root  wheel  224 Feb 20  2024 MacOSX14.4.sdk
drwxr-xr-x   7 root  wheel  224 Apr 10 23:32 MacOSX14.5.sdk
lrwxr-xr-x   1 root  wheel   14 Apr 10 23:31 MacOSX14.sdk -> MacOSX14.5.sdk
drwxr-xr-x   7 root  wheel  224 Mar 18 03:10 MacOSX15.4.sdk
lrwxr-xr-x   1 root  wheel   14 Apr 10 23:31 MacOSX15.sdk -> MacOSX15.4.sdk
```

```
% brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: No Cask quarantine support available: unknown reason.
```

After:
```
ls -al /Library/Developer/CommandLineTools/SDKs
total 0
drwxr-xr-x  8 root  wheel  256 May 12 19:20 .
drwxr-xr-x  5 root  wheel  160 May 12 19:17 ..
lrwxr-xr-x  1 root  wheel   14 May 12 19:20 MacOSX.sdk -> MacOSX15.4.sdk
drwxr-xr-x  7 root  wheel  224 May 12 19:20 MacOSX14.5.sdk
lrwxr-xr-x  1 root  wheel   14 May 12 19:19 MacOSX14.sdk -> MacOSX14.5.sdk
drwxr-xr-x  7 root  wheel  224 Nov 21 09:05 MacOSX15.2.sdk
drwxr-xr-x  7 root  wheel  224 Mar 18 03:10 MacOSX15.4.sdk
lrwxr-xr-x  1 root  wheel   14 May 12 19:19 MacOSX15.sdk -> MacOSX15.4.sdk
```

```
% brew doctor                      
Your system is ready to brew.
```

**Comment by patricklucas** [comment]
Edit: this was a temporary issue on the Apple Developer website and is no longer happening.

Sorry to resurrect this, but it's the most recent discussion/issue I can find.

I'm on Sonoma, still supported by Homebrew, and getting this error. I tried the solutions of deleting the `CommandLineTools` dir and reinstalling with `xcode-select --install`, but it restored it to the same state, with 13.3, 14.5, and 15.2, and I still get the same error when updating casks or running `brew doctor`.

Additionally, the link everyone shares, to https://developer.apple.com/download/all/, just redirects to https://developer.apple.com/account for me after logging in. If I navigate to the .dmg URL directly, I get an error page with:

> Your account can’t access this page.
> There may be certain requirements to view this content.

Any ideas?

**Reply by lzandman** [reply]
Did you log in with your Apple account? Because that's required. However, I don't think you need an active Apple Developer subscription. Mine's expired and I can still access these downloads.

All Xcode Commandline Tools downloads are here: https://developer.apple.com/download/all/?q=Command%20Line%20Tools
If that link doesn't work/redirects, you can always try to manually navigate from that main landing page to the downloads section. Scroll down and click the blue arrow "Get the latest tools and beta SDKs" link. Then under "Development essentials" "Command Line Tools" click the "View downloads" link and you should be on that download page.

**Reply by patricklucas** [reply]
@lzandman yes, this behavior is when I'm logged in. If I visit any of these URLs without having logged in, I first get a login page.

Actually, it appears to be working now. There may have just been a temporary issue on Apple's end (I tried multiple devices/browsers/private window yesterday, and even the link to `/all` from the Apple Developer page resulted in this redirect).
